### PR TITLE
template for algoeditor

### PR DESCRIPTION
**Pull request by kaydee0502**

### CODE REVIEW

File: devsnest/lib/challenge_templates.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module Templates
+  # enum data_types: %i[:int, ]
+
+  class Python3
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:6]: Inherit from a base class which defines the API

File: devsnest/lib/challenge_templates.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    end
+
+    def tail(tail_code)
+      "if __name__ == '__main__':\n\t#{tail_code.join("\n")}"
+    end
+
+    def add_argument(name, type)
+      @func_arguments << [name, type]
+      @tail << input(type[0], type[1], name)
+    end
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:41]: It's better to pull "add_argument" and "define_output" in the base class as they will be common for all

Then, "add_argument" becomes repetitive and better to change it to "add_arguments"

Thinking about it further, you don't need these APIs now. You can just pass these to the build as arguments

File: devsnest/lib/challenge_templates.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        "print(*#{name})"
+      when 'matrix'
+        "for i in #{name}:\n\tprint(*i)"
+      end
+    end
+
+    def build
+      @head = ''
+      @body = "def solve(#{build_argument_list(@func_arguments)}):\n\t# WRITE YOUR CODE HERE."
+      @tail = "#{tail(@tail)}\n#{@output_string}"
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:66]: nit: Not entirely consistent with the rest of the APIs in this class

File: devsnest/lib/challenge_templates.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    def input(type, data_type, name)
+      case type
+      when 'string'
+        "#{name} = input()"
+      when 'array', 'list', 'tuple'
+        "#{name} = #{type}(map(#{data_type}, input().split()))"
+      when 'primitive'
+        "#{name} = #{data_type}(input())"
+      when 'matrix'
+        "n,m = list(map(int, input().split()))\n#{name}=[ [#{data_type}(j) for j in input().split()] for i in range(n)]"
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:30]: This is hard to read. Split into multiple lines

File: devsnest/lib/challenge_templates.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+module Templates
+  # enum data_types: %i[:int, ]
+
+  class Python3
+    # ch = Templates::Python.new
+    # ch.add_argument("n", ["primitive", "int"])
+    # ch.add_argument("arr", ["array", "int"])
+    # ch.define_output("out", "array" )
+    # head, tail, body = ch.build()
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:11]: This comment is redundant when moved to the base class

File: devsnest/lib/challenge_templates.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module Templates
+  # enum data_types: %i[:int, ]
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:4]: Define in a common place and no need to comment

File: devsnest/lib/challenge_templates.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    def define_output(name, type)
+      @output_string = "#{name} = solve(#{build_argument_list(@func_arguments)})\n#{output_code(type, name)}"
+    end
+
+    def output_code(type, name)
+      case type
+      when 'string', 'primitive'
+        "print(#{name})"
+      when 'array', 'list', 'tuple'
+        "print(*#{name})"
```
Comment (layog) [inline devsnest/lib/challenge_templates.rb:57]: Make it a loop

File: devsnest/db/schema.rb [hunk 8]
```diff
@@ -29,6 +29,17 @@
     t.index ["user_id", "challenge_id"], name: "index_algo_submissions_on_user_id_and_challenge_id"
   end
 
+  create_table "algo_templates", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb3", force: :cascade do |t|
+    t.integer "challenge_id"
+    t.string "head"
```
[the remark this answers is not in the document]
Comment (kaydee0502) [reply devsnest/db/schema.rb:34]: > Can you confirm string is free form DS and does not have a character limit?

limit is 255 chars, better change it to :text?

File: devsnest/db/seeds/algo_challenges_seed.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+output_format: {
+0=>  
+    {
+    "name": "out",
+    "variable": {
+        "datastructure": "array",
+        "dtype": "int"
+    }
+}
+})
```
Comment (vedansh) [inline devsnest/db/seeds/algo_challenges_seed.rb:52]: new line

**Review by LakshitXD — APPROVED** [review]

**Review by Yug-gurnani — APPROVED** [review]

**Pull request merged**